applications/ConstitutiveLawsApplication/custom_constitutive/traction_separation_law.cpp

```diff
@@ -0,0 +1,751 @@
+    for (IndexType i=0; i < p_constitutive_law_vector.size()-1; ++i) {
+            mThresholdModeOne[i] = rMaterialProperties[INTERFACIAL_NORMAL_STRENGTH];
+        }
+    
+    mThresholdModeTwo.resize(p_constitutive_law_vector.size()-1, false);
+    for (IndexType i=0; i < p_constitutive_law_vector.size()-1; ++i) {
+            mThresholdModeTwo[i] = rMaterialProperties[INTERFACIAL_SHEAR_STRENGTH];
+        }
+     
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<unsigned int TDim>
+void  TractionSeparationLaw3D<TDim>::CalculateMaterialResponsePK1(ConstitutiveLaw::Parameters& rValues)
+{
+    CalculateMaterialResponsePK2(rValues);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<unsigned int TDim>
+void  TractionSeparationLaw3D<TDim>::CalculateMaterialResponsePK2(ConstitutiveLaw::Parameters& rValues)
+{
+    KRATOS_TRY;
+
+    // Get Values to compute the constitutive law:
+    Flags& r_flags = rValues.GetOptions();
+
+    // Previous flags saved
+    const bool flag_strain       = r_flags.Is(BaseType::USE_ELEMENT_PROVIDED_STRAIN);
+    const bool flag_const_tensor = r_flags.Is(BaseType::COMPUTE_CONSTITUTIVE_TENSOR);
+    const bool flag_stress       = r_flags.Is(BaseType::COMPUTE_STRESS);
+
+    const Properties& r_material_properties = rValues.GetMaterialProperties();
+
+    // The deformation gradient
+    if (rValues.IsSetDeterminantF()) {
+        const double determinant_f = rValues.GetDeterminantF();
+        KRATOS_ERROR_IF(determinant_f < 0.0) << "Deformation gradient determinant (detF) < 0.0 : " << determinant_f << std::endl;
+    }
+
+    // All the strains must be the same, therefore we can just simply compute the strain in the first layer
+    if (r_flags.IsNot(BaseType::USE_ELEMENT_PROVIDED_STRAIN)) {
+        this->CalculateGreenLagrangeStrain(rValues);
+        r_flags.Set(BaseType::USE_ELEMENT_PROVIDED_STRAIN, true);
+    }
+
+    // The global strain vector, constant
+    const Vector strain_vector = rValues.GetStrainVector();
+
+    if (r_flags.Is(BaseType::COMPUTE_STRESS)) {
+        // Set new flags
+        r_flags.Set(BaseType::COMPUTE_CONSTITUTIVE_TENSOR, false);
+        r_flags.Set(BaseType::COMPUTE_STRESS, true);
+
+        // Auxiliar stress vector
+        const auto it_prop_begin       = r_material_properties.GetSubProperties().begin();
+        Vector auxiliar_stress_vector  = ZeroVector(VoigtSize);
+        Vector delamination_damage_affected_stress_vector  = ZeroVector(VoigtSize);
+
+        // The rotation matrix
+        BoundedMatrix<double, VoigtSize, VoigtSize> voigt_rotation_matrix;
+
+        auto p_constitutive_law_vector = this->GetConstitutiveLaws(); 
```

Comment:
can you try with `auto& r_p_constitutive_law_vector` it is more efficient since we do not make a copy

Reply:
Done